TTS/tts/layers/delightful_tts/encoders.py

```diff
@@ -0,0 +1,172 @@
+import torch
+import torch.nn as nn
+import torch.functional as F
+
+from TTS.tts.models.delightful_tts import ReferenceEncoderConfig, get_mask_from_lengths, stride_lens
+
+
+class UtteranceLevelProsodyEncoder(nn.Module):
+    def __init__(
+        self,
+        args: ModelArgs,
+    ):
+        super().__init__()
+
+        self.E = args.encoder.n_hidden
+        self.d_q = self.d_k = args.encoder.n_hidden
+        ref_enc_gru_size = args.reference_encoder.ref_enc_gru_size
+        ref_attention_dropout = args.reference_encoder.ref_attention_dropout
+        bottleneck_size = args.reference_encoder.bottleneck_size_u
+
+        self.encoder = ReferenceEncoderConfig(args)
+        self.encoder_prj = nn.Linear(ref_enc_gru_size, self.E // 2)
+        self.stl = STL(args)
+        self.encoder_bottleneck = nn.Linear(self.E, bottleneck_size)
+        self.dropout = nn.Dropout(ref_attention_dropout)
+
+    def forward(self, mels: torch.Tensor, mel_lens: torch.Tensor) -> torch.Tensor:
+        """
+        mels --- [N, Ty/r, n_mels*r], r=1
+        out --- [N, seq_len, E]
+        """
+        _, embedded_prosody, _ = self.encoder(mels, mel_lens)
+
+        # Bottleneck
+        embedded_prosody = self.encoder_prj(embedded_prosody)
+
+        # Style Token
+        out = self.encoder_bottleneck(self.stl(embedded_prosody))
+        out = self.dropout(out)
+
+        out = out.view((-1, 1, out.shape[3]))
+        return out
+
+
+class PhonemeLevelProsodyEncoder(nn.Module):
+    def __init__(self, args):
+        super().__init__()
+
+        self.E = args.encoder.n_hidden
+        self.d_q = self.d_k = args.encoder.n_hidden
+        bottleneck_size = args.reference_encoder.bottleneck_size_p
+        ref_enc_gru_size = args.reference_encoder.ref_enc_gru_size
+
+        self.encoder = ReferenceEncoder(args)
+        self.encoder_prj = nn.Linear(ref_enc_gru_size, args.encoder.n_hidden)
+        self.attention = ConformerMultiHeadedSelfAttention(
+            d_model=args.encoder.n_hidden,
+            num_heads=args.encoder.n_heads,
+            dropout_p=args.encoder.p_dropout,
+        )
+        self.encoder_bottleneck = nn.Linear(args.encoder.n_hidden, bottleneck_size)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        src_mask: torch.Tensor,
+        mels: torch.Tensor,
+        mel_lens: torch.Tensor,
+        encoding: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        x --- [N, seq_len, encoder_embedding_dim]
```

Comment:
These shape def docstrings need also reformatting as the other models to be compatible with our documentation.